fix: line wrapping within inline code

chriskrycho opened the pull request:
- Adds a failing test.
- Fixes the issue by having `gen_code` call `gen_str` with the string it builds, before returning it.

Fixes #129.

Note that the failing test results in a timeout after the panic, rather than failing the test suite:

<details><summary>Test output</summary>

```sh
$ cargo test
    Finished test [unoptimized + debuginfo] target(s) in 0.03s
     Running unittests src/lib.rs (target/debug/deps/dprint_plugin_markdown-bb4603beba64faa3)

running 13 tests
test generation::cmark::parsing::parse_link_reference_definitions::tests::it_finds_link_reference_with_new_line_after ... ok
test generation::cmark::parsing::parse_link_reference_definitions::tests::it_finds_multiple_link_references ... ok
test generation::cmark::parsing::parse_link_reference_definitions::tests::it_finds_link_reference ... ok
test generation::cmark::parsing::parse_link_reference_definitions::tests::it_parses_empty_string ... ok
test configuration::builder::tests::use_markdown_defaults_when_global_not_set ... ok
test configuration::builder::tests::handle_global_config ... ok
test generation::cmark::parsing::parse_image::test::it_should_parse_image ... ok
test generation::metadata::test::it_should_strip_plus_plus_plus_header ... ok
test generation::metadata::test::it_should_strip_yaml_header ... ok
test configuration::builder::tests::check_all_values_set ... ok
test generation::utils::test::it_should_find_list_words ... ok
test generation::utils::test::should_unindent ... ok
test format_text::test::strips_bom ... ok

test result: ok. 13 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/newline_test.rs (target/debug/deps/newline_test-2293c22b8eb4ea3a)

running 4 tests
test test_issue35_convert_two_spaces_end_of_line_to_hard_break ... ok
test test_issue35_ignore_two_spaces_before_hard_break ... ok
test test_issue22_with_carriage_return_line_feeds ... ok
test test_issue26_with_carriage_return_line_feeds ... ok

test result: ok. 4 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/spec_test.rs (target/debug/deps/specs-6c950b6b2c6409a0)

     Running specs::BlockQuotes

test specs::BlockQuotes::BlockQuotes_TextWrap_Always ... (6ms)
  should format ok
  should wrap a line when it exceeds the line width ok
test specs::BlockQuotes::Callouts ... (8ms)
  should format ok
  should format when just a callout ok
  should format when has blank line ok
test specs::BlockQuotes::BlockQuotes_All ... (9ms)
  should format ok
  should not wrap a line when it exceeds the line width and text wrapping is to maintain ok
  should keep blank lines in the middle ok

     Running specs::Code

thread '<unnamed>' panicked at /Users/chris/.cargo/registry/src/index.crates.io-6f17d22bba15001f/dprint-core-0.66.2/src/formatting/printer.rs:623:7:
Debug panic! Found a newline in the string. Before sending the string to the printer it needs to be broken up and the newline sent as a PrintItem::NewLine. `this
out`
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Panic in spec 'should wrap inline code across lines' in ./tests/specs/Code/Code_Inline_TextWrap_Always.txt

thread '<unnamed>' panicked at /Users/chris/.cargo/registry/src/index.crates.io-6f17d22bba15001f/dprint-development-0.10.1/src/spec_helpers.rs:152:27:
called `Result::unwrap()` on an `Err` value: Any { .. }
test specs::Code::Code_Inline ... (14ms)
  should format inline blocks ok
  should properly space when last on line with soft break ok
  should properly space when first line after soft break ok
  should format code inside text parens ok
  should format double backticks as-is when it includes backticks ok
  should leave spaces around backticks inside double backticks ok
test specs::Code::Code_Inline_TextWrap_Always has been running for more than 60 seconds
```

</details>

chriskrycho committed:
Add failing test for line breaking bug in inline code

chriskrycho committed:
Apply text wrap correctly to inline code

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dprint/dprint-plugin-markdown?pullRequest=130) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/dprint/dprint-plugin-markdown?pullRequest=130) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dprint/dprint-plugin-markdown?pullRequest=130) it.</sub>

chriskrycho commented:
Reading the source, it looks like the right move may be to have `gen_code` build a string (as it does currently) and call `gen_str` with it (which it does not do)?

chriskrycho commented:
Friendly ping/nudge here @dsherret! I’m sure you’re busy, but would love to see this upstream.

chriskrycho committed:
Add test for inline code wrapping in block quotes

The behavior at the time of introducing this test is:

- An initial line wrap for a too-long line is correct.
- When the plugin finds inline code wrapping across a line and decides
  to rewrap it, it inserts an additional `>` at the end of the leading
  line.

This is a *new* failure mode introduced by the previous change, which
added the first support for handling wrapping inline code across lines
at all.

chriskrycho commented on tests/specs/Code/Code_Inline_TextWrap_Always.txt:
```diff
@@ -18,3 +18,24 @@ Testing
+[expect]
+Testing `this out`.
+
+!! should rewrap inline code across lines in blockquote !!
+> Testing `this
+> out`.
+
+[expect]
+> Testing `this out`.
+
+!! should rewrap inline code across lines which are too long in blockquote !!
+> This is some long text, which will have a bit of inline code across the `80 character limit` and then more after that.
+
+[expect]
+> This is some long text, which will have a bit of inline code across the `80
+> character limit` and then more after that.
```
Discovered this failure mode while testing this PR against the Rust book, where there are a few of these. Should hopefully have a fix in for this tomorrow morning!

chriskrycho replied:
Leaving some breadcrumbs here for myself, given I do not expect to get back to this till January—I spent some time chasing this (after beating my head against the wall that is getting useful debugging set up for Rust 😑) and concluded *so far* that it is down to something amiss in how `dprint-plugin-markdown` is using the result of parsing from `pulldown-cmark`, but haven’t gotten further than that.

Specifically, it ends up seeing an inline code snippet that wraps across a line in a blockquote as *including* the `>` character. It will correctly *introduce* the wrapping for a too-long line, but is not stable because once it has done so, on the next formatting pass it will then interpret it as part of the inline code instead of as a blockquote marker.

A side observation: this is just a challenging issue. Here is a simple sample that I used for testing `pulldown-cmark` itself locally:

```markdown
> This is some `inline
> code`. And it gets
> *rendered* just fine.
> The problem is thus
> not `pulldown-cmark`
> but something in how
> `dprint` uses it.
```

Notice how GitHub’s (line-wrapping-aware) renderer trips in a different way over the exact same thing, *not* wrapping when it *should*:

> This is some `inline
> code`. And it gets
> *rendered* just fine.
> The problem is thus
> not `pulldown-cmark`
> but something in how
> `dprint` uses it.

dsherret committed:
fix: strip block quote markers from multi-line code spans

The `Code` node's text is taken from the raw file text, so when a code
span spans multiple lines within a block quote it also contains the
block quote markers of the continuation lines. Now that the text is run
through the text builder those markers ended up as words in the output.

This is what made the block quote specs added here fail.

dsherret committed:
fix: wrap text within inline code spans

Keeps the backticks out of the text builder so that they always stay
attached to the code span's text, instead of being able to wrap onto a
line of their own when a separator space is used.

The approach and the `inline code newlines` spec are taken from
https://github.com/dprint/dprint-plugin-markdown/pull/171. The
`lineWidth: 40` override in that PR was left out so the existing specs
in the file keep their expectations, and the wrapping spec was written
for the default line width instead.

Co-authored-by: Bryan A. Jones <2125584+bjones1@users.noreply.github.com>

dsherret committed:
fix: preserve code span content when wrapping

Four issues found while reviewing the previous commits:

- `strip_block_quote_markers` used an unbounded `trim_start`, so a `>` that
  was part of the code span's text was stripped as if it were a marker. A
  marker may only be indented up to three spaces, and the single space that
  follows it is part of the marker.
- `in_block_quote_count` was an `i8`, which overflowed at 128 nested block
  quotes. Block quotes nest arbitrarily, unlike the tables the type was
  originally chosen for.
- Running the text of a code span through `gen_str` collapsed the runs of
  spaces within it, which changes what the code span renders as. Only a run
  of a single space or line ending may be wrapped now, because both of those
  render as one space.
- Wrapping could move a word like `#`, `>` or `---` to the start of a line,
  where it starts a new block instead. `is_list_word` only guarded list
  markers, so add `is_block_start_word` for the rest. This could happen in
  plain text as well, not only in code spans.
